Sources/MessagingInApp/State/InAppMessageManager.swift

```diff
@@ -0,0 +1,98 @@
+import CioInternalCommon
+import Foundation
+
+// sourcery: InjectRegisterShared = "InAppMessageManager"
+// sourcery: InjectSingleton
+/// InAppMessageManager is the main class used to interact and manage the InAppMessage state.
+/// It is responsible for dispatching actions, subscribing to state changes and fetching the current state.
+/// It is also responsible for initializing the store with the required middleware and reducer.
+/// It also makes asynchronous calls convenient by wrapping them in non-async methods.
+public class InAppMessageManager {
+    private let logger: CioInternalCommon.Logger
+    private let store: InAppMessageStore
+
+    var state: InAppMessageState {
+        get async { await store.state }
+    }
+
+    init(threadUtil: ThreadUtil) {
+        // swiftlint:disable todo
+        // TODO: Remove this shared instance and use DI once Gist Logger is removed.
+        let diGraph = DIGraphShared.shared
+        self.logger = diGraph.logger
+        self.store = InAppMessageStore(
+            reducer: inAppMessageReducer(logger: logger),
+            state: InAppMessageState(),
+            // Sequence of middleware is important. The order of middleware is the order they are executed.
+            // userAuthenticationMiddleware must be first middleware to ensure user is authenticated before any other action is taken.
+            // messageEventCallbacksMiddleware and errorReportingMiddleware must be last to ensure they are executed after all other middlewares.
+            middleware: [
+                userAuthenticationMiddleware(),
+                routeMatchingMiddleware(logger: logger),
+                modalMessageDisplayStateMiddleware(logger: logger, threadUtil: threadUtil),
+                messageMetricsMiddleware(logger: logger),
+                messageQueueProcessorMiddleware(logger: logger),
+                // TODO: Pass delegate to this middleware once Gist shared instance is removed.
+                messageEventCallbacksMiddleware(delegate: nil),
+                errorReportingMiddleware(logger: logger)
+            ]
+        )
+        // swiftlint:enable todo
+    }
+
+    /// Fetches current state of the InAppMessage store and calls the completion block with result.
+    @discardableResult
+    func fetchState(_ completion: @escaping (InAppMessageState) -> Void) -> Task<InAppMessageState, Never> {
+        Task {
+            let currentState = await state
+            completion(currentState)
+            return currentState
+        }
+    }
```

Comment:
```suggestion
@discardableResult
func fetchState(completion: ((InAppMessageState) -> Void)? = nil) -> Task<InAppMessageState, Never> {
    Task {
        let currentState = await state
        completion?(currentState)
        return currentState
    }
}
```

shouldn't we make callback optional, to allow the function to be used in both callback-style and pure async/await contexts.? 

would help us with tests too where we can probably just use await instead of waiting for callback?

Reply:
The use case is valid, but we don't need to tweak this method. Luckily, there is already `state` property exposed, which can be accessed as follows:

```
await inAppMessageManager.state
```